app/models/manageiq/providers/lenovo/physical_infra_manager/parser/physical_storage_parser.rb

```diff
@@ -91,13 +88,111 @@ def parse_driver(driver)
+        components.each do |component|
+          component['canisters'].each do |canister|
+            canisters << parse_canister(canister)
+          end
+        end
+        canisters
+      end
+
+      def parse_canister(canister)
+        {
+          :serial_number                => canister['serialNumber'],
+          :name                         => canister['cmmDisplayName'],
+          :position                     => canister['position'],
+          :status                       => canister['status'],
+          :health_state                 => canister['health'],
+          :disk_bus_type                => canister['diskBusType'],
+          :phy_isolation                => canister['phyIsolation'],
+          :controller_redundancy_status => canister['controllerRedundancyStatus'],
+          :disks                        => canister['disks'],
+          :disk_channel                 => canister['diskChannel'],
+          :system_cache_memory          => canister['systemCacheMemory'],
+          :power_state                  => canister['powerState'],
+          :host_ports                   => canister['hostPorts'],
+          :hardware_version             => canister['hardwareVersion'],
+          :computer_system              => {
+            :hardware => {
+              :firmwares     => get_canister_firmwares(canister),
+              :guest_devices => get_canister_guest_devices(canister)
+            }
+          }
+        }
+      end
+
+      def get_canister_firmwares(canister)
+        firmwares = []
+        firmware = canister['firmware']
+
+        if firmware.present?
+          firmwares = [
+            unless firmware['storageControllerCpuType'] == 'Not Present'
+              {
+                :name          => firmware['storageControllerCpuType'],
+                :build         => firmware['bundleVersion'],
+                :version       => firmware['hardwareVersion'],
+                :resource_type => 'Hardware'
+              }
+            end
+          ]
+        end
+
+        firmwares.compact
+      end
+
+      def get_canister_guest_devices(canister)
+        network_info = []
+        canister_network_info = canister['networkPorts']
+
+        if canister_network_info.present?
+          physical_network_ports = get_physical_network_ports(canister)
+
+          network_info = [
+            {
+              :address                => canister_network_info['macAddress'],
+              :device_name            => canister_network_info['name'],
+              :device_type            => 'management',
+              :network                => {
+                :ipaddress   => canister_network_info['ipAddress'],
+                :subnet_mask => canister_network_info['networkMask']
+              },
+              :physical_network_ports => physical_network_ports
+            }
+          ]
+        end
+
+        network_info.compact
+      end
+
+      def get_physical_network_ports(canister)
```

Comment:
Please, move this method to `PhysicalNetworkPortsParser`

Reply:
@douglasgabriel Chaged it.